Automattic/jetpack
# Sync: Full sync the term relationships

This implements syncing of the term relationships in full sync, which currently we don't seem to be doing.

Turns out it is a bit specific use case, because that table the only one that doesn't have a unique ID column. So to be able to still chunk properly like we do for other sync modules, we use 2 columns, and wrap the functionality in a new sync module, `Term_Relationships`.

We're introducing some tests to verify that missing or obsolete term relationships are synced, and add some tests to the full sync to ensure that it's properly syncing all term relationships, and is properly chunking them when there are a lot of them to sync.

#### Changes proposed in this Pull Request:
* Sync: Full sync the term relationships

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Part of the effort to fix term syncing and the overall state of sync.

#### Testing instructions:
* Apply D30623-code to your WP.com sandbox.
* Apply this PR to your local Jetpack dev env.
* Enable `WP_DEBUG_LOG` on your site if it's disabled.
* Silently create a term relationship between a post and a term on your site (see below for how to do that).
* Trigger a full sync.
* Make sure the term relationship got synced to the cache site.
* Verify full and incremental sync for other items still works.
* Try testing with some other use cases you can think of?
* Verify the tests pass.
* Verify there are no errors.

#### Additional testing help
* Silent creation term relationship can be done with this piece of code:
```
global $wpdb;
$wpdb->insert(
	$wpdb->term_relationships,
	array(
		'object_id'        => 1234,
		'term_taxonomy_id' => 123,
		'term_order'       => 0,
	),
	array(
		'%d',
		'%d',
		'%d',
	)
);
```

#### Proposed changelog entry for your changes:
* Sync: Full sync the term relationships
